### PR TITLE
Add Local Mode Support for Composite Transform and Binary Read

### PR DESCRIPTION
**Pull request by karanataryn**
Adds the ability to be able to run most notebooks using `ExecMode.LOCAL`

### CODE REVIEW

File: lib/sycamore/sycamore/connectors/file/file_scan.py [hunk 1]
```diff
@@ -166,6 +165,47 @@ def execute(self, **kwargs) -> "Dataset":
...
+                if self._metadata_provider:
+                    document.properties.update(self._metadata_provider.get_metadata(path))
+
+                documents.append(document)
+            elif not self._filesystem.get_file_info(path).is_file:
+                file_selector = FileSelector(path)
+                for item in self._filesystem.get_file_info(file_selector):
+                    process_path(item.path)
+
+        for path in paths:
```
Comment (eric-anderson) [inline lib/sycamore/sycamore/connectors/file/file_scan.py:204]: You can simplify the code here.

```
for path in paths:
   for f in self._filesystem.get_file_info(FileSelector(path), recursive=True):
        process_file(f)

def process_file(info):
    if not info.is_file:
        return
    if self._filter_paths_by_extension and not info.path.endswith(self.format()):
        return
   with self._filesystem.open_input_file(path) as file:
   ...
```

File: lib/sycamore/sycamore/docset.py [hunk 2]
```diff
@@ -835,7 +835,7 @@ def llm_filter(
         llm: LLM,
         new_field: str,
         prompt: Union[list[dict], str],
-        field: Optional[str] = "text_representation",
```
Comment (eric-anderson) [inline lib/sycamore/sycamore/docset.py:838]: This feels like an unrelated change. It's generally best to have those in separate PRs, but if necessary make sure it's included in the description.

File: lib/sycamore/sycamore/transforms/base.py [hunk 3]
```diff
@@ -297,5 +298,15 @@ def _local_process(self, in_docs: list[Document]) -> list[Document]:
 
         return docs
 
+    def local_execute(self, all_docs: list[Document]) -> list[Document]:
+        docs = [d for d in all_docs if not isinstance(d, MetadataDocument)]
+        metadata = [d for d in all_docs if isinstance(d, MetadataDocument)]
+        outputs = self._local_process(docs)
+        to_docs = [d for d in outputs if not isinstance(d, MetadataDocument)]
+        if self._enable_auto_metadata and (len(docs) > 0 or len(to_docs) > 0):
+            outputs.extend(BaseMapTransform._update_lineage(docs, to_docs))
```
Comment (eric-anderson) [inline lib/sycamore/sycamore/transforms/base.py:307]: This works, but is a bit weird to call the private method on an unrelated class. Please move _update_lineage out of BaseMapTransform.

File: lib/sycamore/sycamore/connectors/file/file_scan.py (hunk 1 above)
Comment (karanataryn) [reply lib/sycamore/sycamore/connectors/file/file_scan.py:204]: Included, thanks

File: lib/sycamore/sycamore/docset.py (hunk 2 above)
Comment (karanataryn) [reply lib/sycamore/sycamore/docset.py]: Moved it to a separate PR: https://github.com/aryn-ai/sycamore/pull/720

File: lib/sycamore/sycamore/connectors/file/file_scan.py [hunk 4]
```diff
@@ -173,37 +173,37 @@ def local_source(self, **kwargs) -> list[Document]:
...
+                return
+
+            with self._filesystem.open_input_file(info.path) as file:
+                binary_data = file.read()
+
+            document = Document()
+            document.doc_id = str(uuid.uuid1())
+            document.type = self._binary_format
+            document.binary_representation = binary_data
+            document.properties.update({"path": info.path})
```
Comment (eric-anderson) [inline lib/sycamore/sycamore/connectors/file/file_scan.py:189]: nit: you can shorten this 
`document.properties["path"] = info.path`

File: lib/sycamore/sycamore/docset.py (hunk 2 above)
Comment (eric-anderson) [reply lib/sycamore/sycamore/docset.py]: It's still showing up in my diff, but maybe it's github diff weirdness.

File: lib/sycamore/sycamore/transforms/base.py [hunk 5]
```diff
@@ -4,6 +4,7 @@
 import numpy as np
 
 from sycamore.data import Document, MetadataDocument
+from sycamore.utils.lineage_utils import _update_lineage
```
Comment (eric-anderson) [inline lib/sycamore/sycamore/transforms/base.py:7]: nit: import update_lineage
(it's a public function of lineage_utils)

**Review by eric-anderson — APPROVED** [review]

File: lib/sycamore/sycamore/docset.py (hunk 2 above)
Comment (karanataryn) [reply lib/sycamore/sycamore/docset.py]: Yeah that must be it, the PR above is merged.
